[Docs] Add Recipes (#8851)
skypilot-org/skypilot: docs/source/reference/recipes.rst

.. _recipes:

SkyPilot Recipes
================

Recipes is a feature in the SkyPilot dashboard for storing and sharing best-practice patterns for working with SkyPilot. With Recipes, you can:

- **Reuse patterns across workflows**: Create recipes for clusters, managed jobs, pools, and volumes
- **Launch quickly**: Start from a copyable CLI command (e.g., ``sky launch recipes:dev-cluster``) - no YAML file required
- **Edit in place**: View and modify recipes directly in the dashboard
- **Share templates across your organization**: Customize your YAMLs once, then share institutional setup across your team

.. tip::

>    Recipes requires the SkyPilot API server and dashboard. See :ref:`API Server Setup <sky-api-server>` to get started.

Reviewer: Can recipes be used with local API server? If so, remove?
```suggestion
```

Getting started
---------------

Recipes are managed through the SkyPilot dashboard. To access Recipes:

1. Start the API server: ``sky api start``

> 2. Open the dashboard at ``http://localhost:46580`` (or your configured URL)

Reviewer: Can condense these two steps into one:
```suggestion
1. Run `sky dashboard`
```

3. Navigate to the **Recipes** section in the navbar

.. image:: /images/recipes-page.png
   :alt: SkyPilot Recipes page in the dashboard
   :align: center

> Recipes support these SkyPilot resource types: **Clusters**, **Managed Jobs**, **Pools**, and **Volumes**.

Reviewer: Hmm this appears to be too close to the figure. Any way to add a gap/margin?

<img width="831" height="524" alt="Image" src="https://github.com/user-attachments/assets/472e0bd0-a252-404c-b8e4-e7ec59d65ec5" />

Creating a recipe
-----------------

To create a new recipe in the dashboard:

1. Click **New Recipe** in the Recipes page
2. Enter a name and description for your recipe
3. Select the resource type (cluster, managed job, pool, or volume)
4. Write or paste your YAML configuration
5. Click **Create Recipe** to store the recipe

.. image:: /images/recipes-create-form.png
   :alt: Create recipe form in the dashboard
   :align: center

.. note::
   **File and Code Storage Limitations**

   Recipes do not support local file uploads. For code and data:

   - **Code**: Use Git repositories with ``workdir`` dictionary format:

     .. code-block:: yaml

        workdir:
          url: https://github.com/org/repo

>    - **Data**: Use cloud storage buckets with ``file_mounts`` (e.g., ``s3://bucket``, ``gs://bucket``)

Reviewer: Can we suggest volumes too? Also can link to the respective docs pages.

- **YAML Configuration**: The full task specification
- **Launch Command**: Copy-paste command to launch the recipe
- **Metadata**: Name, description, type, creator, and update time

.. image:: /images/recipes-detail-view.png
   :alt: Recipe detail view showing YAML and launch command
   :align: center

To **edit** a recipe, open it and click **Edit**, then update the YAML, name, or description. Changes are saved immediately and available to all users.

To **delete** a recipe, open it and click **Delete**, then confirm. Note that deleting a recipe is permanent and cannot be undone. Existing clusters or jobs launched from the recipe will continue running.

Launching from recipes
----------------------

The most convenient way to use recipes is through the CLI with the ``recipes:`` prefix:

.. code-block:: bash

   # Launch a cluster from a recipe
   sky launch recipes:dev-cluster

   # Launch with custom cluster name
   sky launch -c my-dev recipes:dev-cluster

   # Launch a managed job from a recipe
   sky jobs launch recipes:training-job

   # Launch with overrides
   sky launch recipes:gpu-cluster --cpus 16 --gpus V100:4

> No YAML file needed - the configuration is fetched directly from the dashboard!

Reviewer: ```suggestion
No YAML file needed - the configuration is fetched directly from the SkyPilot API Server!
```

.. image:: /images/recipes-launch-command.png
   :alt: Copy launch command from recipe detail page
   :align: center

> You can also launch directly from the dashboard by clicking **Launch Recipe**, customizing parameters if needed, and clicking **Launch**.

Reviewer: Is this true? I think this is not supported yet

Sharing recipes across your team
---------------------------------

> One of the key benefits of Recipes is enabling teams to share standardized configurations. Here are common patterns with concrete examples:

Reviewer: ```suggestion
A key benefit of Recipes is enabling teams to share standardized configurations. Here are common patterns with concrete examples:
```

**Development Environment** - Standard onboarding setup for new team members

> .. code-block:: yaml

Reviewer: Should we put these examples in a tab set?

     pip install ipython jupyter black pytest

   run: |
     echo "Development environment ready!"
     echo "Connect with: ssh dev-env"

**Use case:** New engineers can get started immediately: ``sky launch recipes:dev-env``

**GPU Training Cluster** - Production training configuration with A100 GPUs

.. code-block:: yaml

   name: gpu-training

   resources:
     accelerators: A100:8
     memory: 128+
     disk_size: 500

   setup: |
     conda create -n train python=3.10 -y
     conda activate train
     pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu121
     pip install transformers datasets wandb

   run: |
     nvidia-smi
     python -c "import torch; print(f'CUDA available: {torch.cuda.is_available()}')"

**Use case:** Everyone uses the same training configuration: ``sky launch recipes:gpu-training``

**Batch Processing Job** - Cost-optimized data processing on spot instances

.. code-block:: yaml

   name: batch-process

   resources:
     cpus: 16+
     memory: 64+
     use_spot: true

   file_mounts:
     /data:
       source: s3://my-data-bucket
       mode: MOUNT
     /output:
       source: s3://my-output-bucket

>        mode: MOUNT

Reviewer: Can also show a volume mount

   setup: |
     pip install pandas pyarrow

   run: |
     python process_data.py --input /data --output /output

**Use case:** Standardized batch processing with automatic recovery: ``sky jobs launch recipes:batch-process``

**Jupyter Notebook Server** - Shared data science environment

.. code-block:: yaml

   name: jupyter-server

   resources:
     cpus: 8+
     memory: 32+
     disk_size: 100
     ports: 8888

   setup: |
     conda create -n jupyter python=3.10 -y
     conda activate jupyter
     pip install jupyterlab pandas numpy scikit-learn matplotlib seaborn

   run: |
     conda activate jupyter
     jupyter lab --ip=0.0.0.0 --port=8888 --no-browser --allow-root

**Use case:** Team data exploration with consistent libraries: ``sky launch recipes:jupyter-server``

**Tips for team collaboration:**

- Use descriptive names: ``prod-inference-v2``, ``staging-training``, ``dev-jupyter``
- Document requirements in the recipe description (setup time, cost estimates, environment variables)

> - Version your infrastructure with naming conventions

Reviewer: Not very clear what this means, can skip

Best practices
--------------

**Naming:** Use clear, descriptive names like ``dev-cluster-python311``, ``training-a100-8gpu-pytorch``, or ``inference-server-vllm``. Avoid generic names like ``test`` or ``recipe1``. Include environment or version information when relevant (``prod-inference-v2``, ``staging-training``).

**Documentation:** Always include a description explaining what the recipe is for, when to use it, who the intended user is, and any special requirements. Example: *"GPU training cluster with PyTorch 2.0 and 8x A100 GPUs. Use for large model training. Requires S3 data bucket configured. Costs ~$20/hour."*

**Configuration tips:**

- Use environment variables for flexibility:

  .. code-block:: yaml

     run: |
       python train.py --data $DATA_PATH --epochs $EPOCHS

  Then customize at launch: ``sky launch recipes:training --env DATA_PATH=s3://my-data``

- Separate ``setup`` (one-time installation) from ``run`` (actual workload) for reusability
- Use ``workdir`` with Git repositories for code syncing:

  .. code-block:: yaml

     workdir:
       url: https://github.com/org/repo

> - Use cloud storage buckets (S3, GCS) for ``file_mounts`` - local file uploads are not supported

Reviewer: Can remove this, already mentioned above

- Be specific about resource requirements (``cpus: 8+``, ``memory: 32+``, ``disk_size: 100``)

> - Consider cost optimization with ``use_spot: true`` and ``any_of`` for GPU fallbacks

Reviewer: These are fairly generic, can be skipped
